### PR TITLE
Future

### PR DESCRIPTION
**Pull request by francescoalemanno**
The aim of this PR is completely refactoring all algorithms into a single powerful algorithm with multiple available moves,
inspired by Differential Evolution MCMC, Affine Invariant MCMC.
whether this will be merged in full, in part, or not is still to be decided.

### CODE REVIEW

**Comment by francescoalemanno** [comment]
@jbrea what do you think about this?

File: src/types.jl [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    rand(rng, density.prior)
+length(density::AbstractApproxPosterior) = length(density.prior)
+
+
+densitytypes(S::AbstractApproxPosterior) = densitytypes(S.prior)
+densitytypes(S::ContinuousDistribution) = Float64
+densitytypes(S::DiscreteDistribution) = Int64
+
+floatize(S) = floatize.(S)
+floatize(S::Number) = convert(Float64, S)
```
Comment (jbrea) [inline src/types.jl:54]: Why not using built-in `float`?

**Review by jbrea — COMMENTED** [review]
Cool, nice work!

So far I went only very quickly over this. I like the idea of making it more general based on the interface functions `accept` and `loglike`. But I would need to think a bit more about the definitions of the structures, e.g. putting `costs` into `ApproximatePosterior`s didn't seem too intuitive to me at first glance and maybe it would make sense to define some `struct`s for the particles?

Also, shall we look into AbstractMCMC.jl as you mentioned and was brought up again in JuliaApproxInference/ApproxInferenceBase.jl#1?

Please let me know if you want me to look in more details at this PR or something in particular.

**Comment by francescoalemanno** [comment]
I agree on both points, a `Particle` struct Is a good idea, and It would be very nice if both my affine_mcmc and your adaptive_smc become implementations of `AbstractMCMC`, altough at the moment I find it difficult to follow the logic of that abstract interface.. 🤣

in this case, i will merge this, and slowly work on another PR to include a `Particle` struct

**Pull request merged**